packages/services/src/kernel/manager.ts

```diff
@@ -195,71 +195,126 @@ export class KernelManager implements Kernel.IManager {
    * This will emit [[runningChanged]] if the running kernels list
    * changes.
    */
-  shutdown(id: string): Promise<void> {
-    let index = ArrayExt.findFirstIndex(this._models, value => value.id === id);
+  async shutdown(id: string): Promise<void> {
+    const models = this._models;
+    const kernels = this._kernels;
+    const index = ArrayExt.findFirstIndex(models, value => value.id === id);
+
     if (index === -1) {
       return;
     }
 
     // Proactively remove the model.
-    this._models.splice(index, 1);
-    this._runningChanged.emit(this._models.slice());
-
-    return Kernel.shutdown(id, this.serverSettings).then(() => {
-      let toRemove: Kernel.IKernel[] = [];
-      this._kernels.forEach(k => {
-        if (k.id === id) {
-          k.dispose();
-          toRemove.push(k);
-        }
-      });
-      toRemove.forEach(k => {
-        this._kernels.delete(k);
-      });
+    models.splice(index, 1);
+
+    // Delete and dispose the kernel locally.
+    kernels.forEach(kernel => {
+      if (kernel.id === id) {
+        kernels.delete(kernel);
+        kernel.dispose();
+      }
     });
+
+    // Shut down the remote session.
+    await Kernel.shutdown(id, this.serverSettings);
+
+    // Emit the new model list.
+    this._runningChanged.emit(models.slice());
```

Comment:
Note to self: before we emitted runningChange before shutdown and disposed clientside kernels after shutdown. Now we dispose clientside kernels before shutdown and emit runningChanged after.

Reply:
Resolved in afshin#24